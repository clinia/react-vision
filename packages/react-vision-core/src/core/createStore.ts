type SearchState = any;
type ResultsState = any;
type SuggestionsState = any;
type Listener = () => void;
type State = {
  widgets: SearchState;
  metadata: any[];
  results: ResultsState | null;
  resultsSuggestions: SuggestionsState | null;
  error: Error | null;
  searching: boolean;
  searchingForSuggestions: boolean;
  isSearchStalled: boolean;
};
export default function createStore(initialState: State) {
  let state = initialState;
  const listeners: Listener[] = [];
  return {
    getState() {
      return state;
    },
    setState(nextState: State) {
      state = nextState;
      listeners.forEach(listener => listener());
    },
    subscribe(listener: Listener) {
      listeners.push(listener);
      return function unsubscribe() {
        listeners.splice(listeners.indexOf(listener), 1);
      };
    },
  };
}

export type Store = ReturnType<typeof createStore>;
