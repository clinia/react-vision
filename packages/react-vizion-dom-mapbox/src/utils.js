import PropTypes from 'prop-types';

export const registerEvents = (events, props, instance) => {
  const eventsAvailable = Object.keys(events);
  const keys = Object.keys(props).filter(
    key => eventsAvailable.indexOf(key) !== -1
  );
  keys.forEach(name =>
    instance.addEventListener(events[name], event => {
      props[name]({ event, marker: instance });
    })
  );

  return () => {
    keys.forEach(name =>
      instance.removeEventListener(events[name], event => {
        props[name]({ event, marker: instance });
      })
    );
  };
};

export const createListenersPropTypes = eventTypes =>
  Object.keys(eventTypes).reduce(
    (acc, name) => ({ ...acc, [name]: PropTypes.func }),
    {}
  );

export const createFilterProps = excludes => props =>
  Object.keys(props)
    .filter(name => excludes.indexOf(name) === -1)
    .reduce(
      (acc, name) => ({
        ...acc,
        [name]: props[name],
      }),
      {}
    );

export const betweenRange = (value, n1, n2) => {
  return n1 <= value && n2 >= value;
};

export const isValidCoordinates = (lat, lon) => {
  return latitude(lat) && longitude(lon);
};

const isNumber = n => {
  return typeof n === 'number';
};

function longitude(lon) {
  return Boolean(isNumber(lon) && betweenRange(lon, -180, 180));
}

function latitude(lat) {
  return Boolean(isNumber(lat) && betweenRange(lat, -90, 90));
}
