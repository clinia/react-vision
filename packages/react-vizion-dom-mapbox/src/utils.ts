export const betweenRange = (value: number, n1: number, n2: number) => {
  return n1 <= value && n2 >= value;
};

export const isValidCoordinates = (lat: any, lon: any) => {
  return latitude(lat) && longitude(lon);
};

const isNumber = (n: any) => {
  return typeof n === 'number';
};

function longitude(lon: any) {
  return Boolean(isNumber(lon) && betweenRange(lon, -180, 180));
}

function latitude(lat: any) {
  return Boolean(isNumber(lat) && betweenRange(lat, -90, 90));
}
