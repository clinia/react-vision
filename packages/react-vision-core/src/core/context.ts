import { createContext } from 'react';
import { Store } from '../core/createStore';

export type VisionContext = {
  onInternalStateUpdate: (...args: any[]) => any;
  createHrefForState: (...args: any[]) => string;
  onSearchForSuggestions: (...args: any[]) => any;
  onSearchStateChange: (...args: any[]) => any;
  onSearchParameters: (...args: any[]) => any;
  store: Store;
  widgetsManager: any;
  mainTargetedIndex: string;
};

export const {
  Consumer: VisionConsumer,
  Provider: VisionProvider,
} = createContext<VisionContext>({
  onInternalStateUpdate: () => undefined,
  createHrefForState: () => '#',
  onSearchForSuggestions: () => undefined,
  onSearchStateChange: () => undefined,
  onSearchParameters: () => undefined,
  store: {} as Store,
  widgetsManager: {},
  mainTargetedIndex: '',
});

export type IndexContext =
  | {
      targetedIndex: string;
    }
  | undefined;

export const {
  Consumer: IndexConsumer,
  Provider: IndexProvider,
} = createContext<IndexContext>(undefined);
