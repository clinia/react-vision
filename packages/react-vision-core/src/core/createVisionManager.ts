import cliniasearchHelper from 'cliniasearch-helper';
import createWidgetsManager from './createWidgetsManager';
import createStore from './createStore';
import { HIGHLIGHT_TAGS } from './highlight';
import { hasMultipleIndices } from './indexUtils';
import { version as ReactVersion } from 'react';
import version from './version';

function addCliniaAgents(searchClient) {
  if (typeof searchClient.addCliniaAgent === 'function') {
    searchClient.addCliniaAgent(`react (${ReactVersion})`);
    searchClient.addCliniaAgent(`react-vision (${version})`);
  }
}

const isMultiIndexContext = widget =>
  hasMultipleIndices({
    vci: widget.props.contextValue,
    multiIndexContext: widget.props.indexContextValue,
  });
const isTargetedIndexEqualIndex = (widget, indexId) =>
  widget.props.indexContextValue.targetedIndex === indexId;

const isIndexWidget = widget => Boolean(widget.props.indexId);
const isIndexWidgetEqualIndex = (widget, indexId) =>
  widget.props.indexId === indexId;

const sortIndexWidgetsFirst = (firstWidget, secondWidget) => {
  if (isIndexWidget(firstWidget)) {
    return -1;
  }
  if (isIndexWidget(secondWidget)) {
    return 1;
  }
  return 0;
};

/**
 * Creates a new instance of the VisionManager which controls the widgets and
 * trigger the search when the widgets are updated.
 * @param {string} indexName - the main index name
 * @param {object} initialState - initial widget state
 * @param {object} SearchParameters - optional additional parameters to send to the clinia API
 * @param {number} stalledSearchDelay - time (in ms) after the search is stalled
 * @return {VisionManager} a new instance of VisionManager
 */
export default function createVisionManager({
  indexName,
  initialState = {},
  searchClient,
  resultsState,
  stalledSearchDelay,
}) {
  const helper = cliniasearchHelper(searchClient, indexName, {
    ...HIGHLIGHT_TAGS,
  });

  addCliniaAgents(searchClient);

  helper
    .on('search', handleNewSearch)
    .on('result', handleSearchSuccess({ indexId: indexName }))
    .on('error', handleSearchError);

  let locationClient = searchClient.initPlaces();

  let skip = false;
  let stalledSearchTimer: NodeJS.Timeout | null = null;
  let initialSearchParameters = helper.state;

  const widgetsManager = createWidgetsManager(onWidgetsUpdate);

  hydrateSearchClient(searchClient, resultsState);

  const store = createStore({
    widgets: initialState,
    metadata: [],
    results: hydrateResultsState(resultsState),
    resultsSuggestions: null,
    resultsLocations: null,
    error: null,
    searching: false,
    searchingForSuggestions: false,
    searchingForLocations: false,
    isSearchStalled: true,
  });

  function skipSearch() {
    skip = true;
  }

  function updateClient(client) {
    addCliniaAgents(client);
    helper.setClient(client);
    search();
  }

  function clearCache() {
    helper.clearCache();
    search();
  }

  function getMetadata(state) {
    return widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getMetadata))
      .map(widget => widget.getMetadata(state));
  }

  function getSearchParameters() {
    const sharedParameters = widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getSearchParameters))
      .filter(widget => !isMultiIndexContext(widget) && !isIndexWidget(widget))
      .reduce(
        (res, widget) => widget.getSearchParameters(res),
        initialSearchParameters
      );

    const mainParameters = widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getSearchParameters))
      .filter(widget => {
        const targetedIndexEqualMainIndex =
          isMultiIndexContext(widget) &&
          isTargetedIndexEqualIndex(widget, indexName);

        const subIndexEqualMainIndex =
          isIndexWidget(widget) && isIndexWidgetEqualIndex(widget, indexName);

        return targetedIndexEqualMainIndex || subIndexEqualMainIndex;
      })
      .sort(sortIndexWidgetsFirst)
      .reduce(
        (res, widget) => widget.getSearchParameters(res),
        sharedParameters
      );

    const derivedIndices = widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getSearchParameters))
      .filter(widget => {
        const targetedIndexNotEqualMainIndex =
          isMultiIndexContext(widget) &&
          !isTargetedIndexEqualIndex(widget, indexName);

        const subIndexNotEqualMainIndex =
          isIndexWidget(widget) && !isIndexWidgetEqualIndex(widget, indexName);

        return targetedIndexNotEqualMainIndex || subIndexNotEqualMainIndex;
      })
      .sort(sortIndexWidgetsFirst)
      .reduce((indices, widget) => {
        const indexId = isMultiIndexContext(widget)
          ? widget.props.indexContextValue.targetedIndex
          : widget.props.indexId;

        const widgets = indices[indexId] || [];

        return { ...indices, [indexId]: widgets.concat(widget) };
      }, {});

    const derivedParameters = Object.keys(derivedIndices).map(indexId => ({
      parameters: derivedIndices[indexId].reduce(
        (res, widget) => widget.getSearchParameters(res),
        sharedParameters
      ),
      indexId,
    }));

    return { mainParameters, derivedParameters };
  }

  function onSearchForSuggestions({ query, args }) {
    store.setState({
      ...store.getState(),
      searchingForSuggestions: true,
    });

    searchClient
      .suggest(query, args)
      .then(content => {
        store.setState({
          ...store.getState(),
          searchingForSuggestions: false,
          resultsSuggestions: {
            suggestions: content,
            query,
          },
        });
      })
      .catch(error => {
        // Since setState is synchronous, any error that occurs in the render of a
        // component will be swallowed by this promise.
        // This is a trick to make the error show up correctly in the console.
        // See http://stackoverflow.com/a/30741722/969302
        setTimeout(() => {
          throw error;
        });
      });
  }

  function onSearchForLocations({ query, ...args }) {
    store.setState({
      ...store.getState(),
      searchingForLocations: true,
    });

    if (locationClient) {
      locationClient!
        .suggest(query, args)
        .then(content => {
          store.setState({
            ...store.getState(),
            searchingForLocations: false,
            resultsLocations: {
              suggestions: content,
              query,
            },
          });
        })
        .catch(error => {
          // Since setState is synchronous, any error that occurs in the render of a
          // component will be swallowed by this promise.
          // This is a trick to make the error show up correctly in the console.
          // See http://stackoverflow.com/a/30741722/969302
          setTimeout(() => {
            throw error;
          });
        });
    }
  }

  function search() {
    if (skip) return;

    const { mainParameters, derivedParameters } = getSearchParameters();

    helper.derivedHelpers.slice().forEach(derivedHelper => {
      derivedHelper.detach();
    });

    derivedParameters.forEach(({ indexId, parameters }) => {
      const derivedHelper = helper.derive(() => parameters);

      derivedHelper
        .on('result', handleSearchSuccess({ indexId }))
        .on('error', handleSearchError);
    });

    helper.setState(mainParameters);

    helper.search();
  }

  function handleSearchSuccess({ indexId }) {
    return event => {
      const state = store.getState();
      const isDerivedHelperEmpty = !helper.derivedHelpers.length;

      let results = state.results ? state.results : {};

      if (!isDerivedHelperEmpty) {
        results[indexId] = event;
      } else {
        results = event;
      }

      const currentState = store.getState();
      let nextIsSearchStalled = currentState.isSearchStalled;
      if (!helper.hasPendingRequests()) {
        clearTimeout(stalledSearchTimer!);
        stalledSearchTimer = null;
        nextIsSearchStalled = false;
      }

      const { ...partialState } = currentState;

      store.setState({
        ...partialState,
        results,
        isSearchStalled: nextIsSearchStalled,
        searching: false,
        error: null,
      });
    };
  }

  function handleSearchError({ error }) {
    const currentState = store.getState();

    let nextIsSearchStalled = currentState.isSearchStalled;
    if (!helper.hasPendingRequests()) {
      clearTimeout(stalledSearchTimer!);
      nextIsSearchStalled = false;
    }

    const { ...partialState } = currentState;

    store.setState({
      ...partialState,
      isSearchStalled: nextIsSearchStalled,
      error,
      searching: false,
    });
  }

  function handleNewSearch() {
    if (stalledSearchTimer) return;

    stalledSearchTimer = setTimeout(() => {
      const { ...partialState } = store.getState();

      store.setState({
        ...partialState,
        isSearchStalled: true,
      });
    }, stalledSearchDelay);
  }

  function hydrateSearchClient(client, results) {
    if (!results) return;

    if (!client._useCache || typeof client.addCliniaAgent !== 'function') {
      // Avoid hydrating the client when cache is disable or the client is not ours
      return;
    }

    if (Array.isArray(results)) {
      hydrateSearchClientWithMultiIndexRequest(client, results);
      return;
    }

    hydrateSearchClientWithSingleIndexRequest(client, results);
  }

  function hydrateSearchClientWithMultiIndexRequest(client, results) {
    const key = `/1/indexes/*/queries_body_${JSON.stringify({
      requests: results.reduce((acc, result) =>
        acc.concat(
          result.rawResults.map(request => ({
            indexName: request.index,
            params: request.params,
          }))
        )
      ),
    })}`;

    client.cache = {
      ...client.cache,
      [key]: JSON.stringify({
        results: results.reduce(
          (acc, result) => acc.concat(result.rawResults),
          []
        ),
      }),
    };
  }

  function hydrateSearchClientWithSingleIndexRequest(client, results) {
    const key = `/1/indexes/*/queries_body_${JSON.stringify({
      requests: results.rawResults.map(request => ({
        indexName: request.index,
        params: request.params,
      })),
    })}`;

    client.cache = {
      ...client.cache,
      [key]: JSON.stringify({
        results: results.rawResults,
      }),
    };
  }

  function hydrateResultsState(results) {
    if (!results) return null;

    if (Array.isArray(results)) {
      return results.reduce(
        (acc, result) => ({
          ...acc,
          [result._internalIndexId]: new cliniasearchHelper.SearchResults(
            new cliniasearchHelper.SearchParameters(result.state),
            result.rawResults
          ),
        }),
        {}
      );
    }

    return new cliniasearchHelper.SearchResults(
      new cliniasearchHelper.SearchParameters(results.state),
      results.rawResults
    );
  }

  // Called whenever a widget has been rendered with new props.
  function onWidgetsUpdate() {
    const metadata = getMetadata(store.getState().widgets);

    store.setState({
      ...store.getState(),
      metadata,
      searching: true,
    });

    // Since the `getSearchParameters` method of widgets also depends on props,
    // the result search parameters might have changed.
    search();
  }

  function transitionState(nextSearchState) {
    const searchState = store.getState().widgets;

    return widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.transitionState))
      .reduce(
        (res, widget) => widget.transitionState(searchState, res),
        nextSearchState
      );
  }

  function onExternalStateUpdate(nextSearchState) {
    const metadata = getMetadata(nextSearchState);

    store.setState({
      ...store.getState(),
      widgets: nextSearchState,
      metadata,
      searching: true,
    });

    search();
  }

  function updateIndex(newIndex) {
    initialSearchParameters = initialSearchParameters.setIndex(newIndex);
    // No need to trigger a new search here as the widgets will also update and trigger it if needed.
  }

  function getWidgetsIds() {
    return store
      .getState()
      .metadata.reduce(
        (res, meta) =>
          typeof meta.id !== 'undefined' ? res.concat(meta.id) : res,
        []
      );
  }

  return {
    store,
    widgetsManager,
    getWidgetsIds,
    getSearchParameters,
    onSearchForSuggestions,
    onSearchForLocations,
    onExternalStateUpdate,
    transitionState,
    updateClient,
    updateIndex,
    clearCache,
    skipSearch,
  };
}

export type VisionManager = ReturnType<typeof createVisionManager>;
