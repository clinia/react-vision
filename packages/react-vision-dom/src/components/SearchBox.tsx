import React, { Component } from 'react';
import LoadingIndicator from './LoadingIndicator';
import { createClassNames } from '../core/utils';
import classnames from 'classnames';
import { translatable } from 'react-vision-core';

type Props = {
  currentRefinement?: string;
  searchAsYouType?: boolean;
  className?: string;
  placeholder?: string;
  style?: React.CSSProperties;

  autoFocus?: boolean;
  loading?: boolean;
  showLoadingIndicator?: boolean;
  disabled?: boolean;

  onClear?: (event: any) => void;
  onFocus?: (event: React.FocusEvent) => void;
  onBlur?: (event: React.FocusEvent) => void;
  onInput?: (event: React.FormEvent<HTMLInputElement>) => void;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onSubmit?: (event: any) => void;

  clearButton?: (clearSearch: () => void) => React.ReactNode;
  submitButton?: React.ReactNode;
  loadingIndicator?: React.ReactNode;
  translate: any;
};

type State = {
  query: string;
};

const cx = createClassNames('SearchBox');

class SearchBox extends Component<Props, State> {
  formRef;
  state = {
    query: '',
  };

  constructor(props) {
    super(props);
    // We bind functions for test purposes instead of using arrow functions
    this.onSubmit = this.onSubmit.bind(this);
    this.onInput = this.onInput.bind(this);
    this.onClear = this.onClear.bind(this);
    this.formRef = React.createRef();
  }

  onSubmit(event) {
    event.preventDefault();
    if (this.props.onSubmit) this.props.onSubmit(event);
  }

  onInput(event: React.ChangeEvent<HTMLInputElement>) {
    const query = event.target.value;
    const { searchAsYouType } = this.props;

    this.setState({ query }, () => {
      if (searchAsYouType) {
        this.formRef.dispatchEvent(new Event('submit'));
      }
    });

    if (this.props.onInput) this.props.onInput(event);
  }

  onClear() {
    const query = '';
    this.setState({ query });
  }

  render() {
    const {
      autoFocus,
      className,
      clearButton,
      disabled,
      loading,
      loadingIndicator,
      showLoadingIndicator,
      onBlur,
      onChange,
      onClear,
      onFocus,
      onKeyPress,
      placeholder,
      style,
      submitButton,
      translate,
    } = this.props;

    const { query } = this.state;

    return (
      <div className={classnames(cx(''), className)} style={style}>
        <form onSubmit={this.onSubmit} ref={ref => (this.formRef = ref)}>
          <div>
            <input
              type="search"
              autoComplete="off"
              autoCorrect="off"
              autoCapitalize="off"
              disabled={disabled}
              spellCheck={false}
              className={cx('input')}
              onInput={this.onInput}
              onChange={onChange}
              onKeyPress={onKeyPress}
              onBlur={onBlur}
              autoFocus={autoFocus}
              onFocus={onFocus}
              value={query}
              placeholder={
                placeholder ? placeholder : translatable('placeholder')
              }
            />
            {showLoadingIndicator &&
              loading &&
              (loadingIndicator ? (
                loadingIndicator
              ) : (
                <LoadingIndicator isLoading={loading} />
              ))}
            {clearButton ? (
              clearButton(this.onClear)
            ) : (
              <button
                className={cx('clear')}
                onClick={onClear ? onClear : this.onClear}
              >
                {translate('clearText')}
              </button>
            )}
            {submitButton ? (
              submitButton
            ) : (
              <button className={cx('submit')} type="submit">
                {translate('searchText')}
              </button>
            )}
          </div>
        </form>
      </div>
    );
  }
}

export default translatable({
  placeholder: 'Default',
  searchText: 'Search',
  clearText: 'Clear',
})(SearchBox);
