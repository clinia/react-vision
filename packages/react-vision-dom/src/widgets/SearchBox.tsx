import { connectSearchBox } from 'react-vision-core';
import SearchBox from '../components/SearchBox';

/**
 * @module SearchBox
 */

/**
 * The SearchBox component displays a search box that lets the user search for a specific query.
 * @alias module:SearchBox
 * @kind widget
 * @prop {boolean} autoFocus=false - Should the search box be focused on render?
 * @prop {string} className - Add a custom CSS class to the SearchBox form container.
 * @prop {node} clear - Change the apparence of the default reset button (cross).
 * @prop {string} clearTitle - The reset button title.
 * @prop {string} defaultRefinement - Provide default refinement value when component is mounted.
 * @prop {string[]} focusShortcuts=['s','/'] - List of keyboard shortcuts that focus the search box. Accepts key names and key codes.
 * @prop {node} loadingIndicator - Change the apparence of the default loading indicator (spinning circle).
 * @prop {function} onClear - Listen to `reset` event sent from the SearchBox form container.
 * @prop {function} onSubmit - Intercept submit event sent from the SearchBox form container.
 * @prop {function} on* - Listen to any events sent from the search input itself.
 * @prop {string} placeholder - The label of the input placeholder.
 * @prop {boolean} searchAsYouType=true - Should we search on every change to the query? If you disable this option, new searches will only be triggered by clicking the search button or by pressing the enter key while the search box is focused.
 * @prop {string} searchTitle - The submit button title.
 * @prop {boolean} showLoadingIndicator=false - Display that the search is loading. This only happens after a certain amount of time to avoid a blinking effect. This timer can be configured with `stalledSearchDelay` props on <InstantSearch>. By default, the value is 200ms.
 * @prop {object} style - Add a custom React.CSSProperties object to SearchBox form container.
 * @prop {node} submit - Change the apparence of the default submit button (magnifying glass).
 * @themekey cvi-SearchBox - The root div of the widget.
 * @themekey cvi-SearchBox-clear - The reset button used to clear the content of the input.
 * @themekey cvi-SearchBox-clearIcon - The default reset icon used inside the reset button.
 * @themekey cvi-SearchBox-form - The wrapping form.
 * @themekey cvi-SearchBox-input - The search input.
 * @themekey cvi-SearchBox-loadingIndicator - The loading indicator container.
 * @themekey cvi-SearchBox-loadingIcon - The default loading icon.
 * @themekey cvi-SearchBox-submit - The submit button.
 * @themekey cvi-SearchBox-submitIcon - The default magnifier icon used with the search input.
 * @example
 * import React from 'react';
 * import cliniasearch from 'cliniasearch/lite';
 * import { Vision, SearchBox } from 'react-vision-dom';
 *
 * const searchClient = cliniasearch(
 *   'latency',
 *   '6be0576ff61c053d5f9a3225e2a90f76'
 * );
 *
 * const App = () => (
 *   <Vision
 *     searchClient={searchClient}
 *   >
 *     <SearchBox />
 *   </Vision>
 * );
 */

export default connectSearchBox(SearchBox);
