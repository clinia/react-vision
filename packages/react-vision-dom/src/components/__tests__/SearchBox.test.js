import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SearchBox from '../SearchBox';
import LoadingIndicator from '../LoadingIndicator';
import { createClassNames } from '../../core/utils';

Enzyme.configure({ adapter: new Adapter() });

/*NOTES:
Tests that uses .dive() do that to get the inner component inside the hoc
To our tests it's necessary to retrieve the inner component to simulate events and etc
*/
describe('SearchBox', () => {
  const cx = createClassNames('SearchBox');
  it('applies its default props', () => {
    const instance = renderer.create(<SearchBox refine={() => null} />);

    expect(instance.toJSON()).toMatchSnapshot();

    instance.unmount();
  });

  it('should have focus after render when autoFocus prop is true', () => {
    //Creates a parent div where the component will be attached to
    //so that way we can get the active element id
    document.body.innerHTML = '<div></div>';

    const wrapper = mount(<SearchBox autoFocus={true} />, {
      attachTo: document.getElementsByName('div')[0],
    });

    const elementInstance = wrapper.find('input').instance();
    expect(elementInstance).toEqual(document.activeElement);
  });

  it('should initialize state with empty query', () => {
    const wrapper = shallow(<SearchBox />).dive();
    expect(wrapper.state('query')).toEqual('');
  });

  it('should render input disabled when "disabled" props is true', () => {
    const wrapper = shallow(<SearchBox disabled={true} />).dive();
    expect(wrapper.find('input').props('disabled')).toBeTruthy();
  });

  describe('Style tests', () => {
    it('should render main div with className props', () => {
      const classNameTest = 'class-test';
      const wrapper = shallow(<SearchBox className={classNameTest} />);

      expect(wrapper.first('div').hasClass(classNameTest)).toBeTruthy();
    });

    it('should render main div with style props ', () => {
      const styleTest = { top: 0 };
      const wrapper = shallow(<SearchBox style={styleTest} />);

      expect(wrapper.first('div').props().style).toEqual(styleTest);
    });
  });

  describe('Custom inner components', () => {
    it('should render with custom loading indicator', () => {
      const customLoadingIndicator = <span>Loading</span>;
      const wrapper = shallow(
        <SearchBox
          loadingIndicator={customLoadingIndicator}
          showLoadingIndicator={true}
          loading={true}
        />
      ).dive();
      const defaultLoadingIndicator = <LoadingIndicator />;

      expect(wrapper.contains(customLoadingIndicator)).toBeTruthy();
      expect(wrapper.contains(defaultLoadingIndicator)).toBeFalsy();
    });

    it('should not render with custom loading indicator when "showLoadingIndicator" props is false', () => {
      const customLoadingIndicator = <span>Loading</span>;
      const wrapper = shallow(
        <SearchBox
          loadingIndicator={customLoadingIndicator}
          showLoadingIndicator={false}
          loading={true}
        />
      );
      const defaultLoadingIndicator = <LoadingIndicator />;

      expect(wrapper.contains(customLoadingIndicator)).toBeFalsy();
      expect(wrapper.contains(defaultLoadingIndicator)).toBeFalsy();
    });

    it('should not render with custom loading indicator when "loading" props is false', () => {
      const customLoadingIndicator = <span>Loading</span>;
      const wrapper = shallow(
        <SearchBox
          loadingIndicator={customLoadingIndicator}
          loading={false}
          showLoadingIndicator={true}
        />
      );
      const defaultLoadingIndicator = <LoadingIndicator />;

      expect(wrapper.contains(customLoadingIndicator)).toBeFalsy();
      expect(wrapper.contains(defaultLoadingIndicator)).toBeFalsy();
    });

    it('should render with custom clear button', () => {
      const customClearButton = <button>Test</button>;
      const wrapper = shallow(
        <SearchBox clearButton={_ => customClearButton} />
      ).dive();

      expect(wrapper.contains(customClearButton)).toBeTruthy();
      expect(wrapper.exists('#search-box-clear')).toBeFalsy();
    });

    it('should render with custom submit button', () => {
      const customSubmitButton = <button>Test</button>;

      const wrapper = shallow(
        <SearchBox submitButton={customSubmitButton} />
      ).dive();

      expect(wrapper.contains(customSubmitButton)).toBeTruthy();
      expect(wrapper.exists('.search-box-button')).toBeFalsy();
    });
  });

  describe('Event tests', () => {
    describe('Default events', () => {
      it('should update state on default input change', () => {
        const wrapper = shallow(<SearchBox />).dive();
        const enteredInputValue = 'a';

        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state('query')).toEqual(enteredInputValue);
      });

      it('should not submit on default input change when "searchAsYouType" prop is not valid ', () => {
        const wrapper = shallow(<SearchBox />).dive();
        const enteredInputValue = 'a';
        const spyOnSubmit = spyOn(wrapper.instance(), 'onSubmit');

        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state().query).toEqual(enteredInputValue);
        expect(spyOnSubmit).not.toHaveBeenCalled();
      });

      it('should submit on default input change when "searchAsYouType" prop is true ', () => {
        //Mounts the inner component inside the hoc
        const wrapper = mount(
          shallow(<SearchBox searchAsYouType={true} />).get(0)
        );
        const enteredInputValue = 'a';
        const spyOnSubmit = spyOn(wrapper.instance(), 'onSubmit');

        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state().query).toEqual(enteredInputValue);
        expect(spyOnSubmit).toHaveBeenCalled();
      });
      it('should call submit on form submit', () => {
        const wrapper = shallow(<SearchBox />).dive();
        const spyOnSubmit = spyOn(wrapper.instance(), 'onSubmit');
        wrapper.instance().forceUpdate();

        wrapper.find('form').simulate('submit', { preventDefault: () => {} });

        expect(spyOnSubmit).toHaveBeenCalled();
      });
      it('should clear state on default clear search click', () => {
        const wrapper = shallow(<SearchBox />).dive();
        const enteredInputValue = 'a';

        //Simulate input to change the default state query value
        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state().query).toEqual(enteredInputValue);

        const clearElement = wrapper.find(`.${cx('clear')}`);
        clearElement.simulate('click');

        expect(wrapper.state().query).toEqual('');
      });
    });

    describe('Custom events', () => {
      it('should trigger onBlur on input blur', () => {
        const onBlur = jest.fn();

        const wrapper = shallow(<SearchBox onBlur={onBlur} />).dive();
        wrapper.find('input').simulate('blur');

        expect(onBlur).toHaveBeenCalled();
      });
      it('should trigger onChange on input change', () => {
        const onChange = jest.fn();

        const wrapper = shallow(<SearchBox onChange={onChange} />).dive();
        wrapper.find('input').simulate('change');

        expect(onChange).toHaveBeenCalled();
      });
      it('should trigger onClear on input change', () => {
        const onClear = jest.fn();

        const wrapper = shallow(<SearchBox onClear={onClear} />).dive();
        const clearElement = wrapper.find(`.${cx('clear')}`);
        clearElement.simulate('click');

        expect(onClear).toHaveBeenCalled();
      });
      it('should trigger onFocus on input focus', () => {
        const onFocus = jest.fn();

        const wrapper = shallow(<SearchBox onFocus={onFocus} />).dive();
        wrapper.find('input').simulate('focus');

        expect(onFocus).toHaveBeenCalled();
      });
      it('should trigger onInput on input change', () => {
        const onInput = jest.fn();
        const enteredInputValue = 'a';
        const wrapper = shallow(<SearchBox onInput={onInput} />).dive();
        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(onInput).toHaveBeenCalled();
      });
      it('should trigger onKeyPress on input change', () => {
        const onKeyPress = jest.fn();

        const wrapper = shallow(<SearchBox onKeyPress={onKeyPress} />).dive();
        wrapper.find('input').simulate('keyPress');

        expect(onKeyPress).toHaveBeenCalled();
      });
      it('should trigger onSubmit on input change', () => {
        const onSubmit = jest.fn();

        const wrapper = shallow(<SearchBox onSubmit={onSubmit} />).dive();
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });

        expect(onSubmit).toHaveBeenCalled();
      });
    });
  });
});
