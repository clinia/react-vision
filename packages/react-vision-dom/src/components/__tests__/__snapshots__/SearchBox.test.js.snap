// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox applies its default props 1`] = `
<div
  className="cvi-SearchBox"
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        className="cvi-SearchBox-input"
        onInput={[Function]}
        placeholder={[Function]}
        spellCheck={false}
        type="search"
        value=""
      />
      <button
        className="cvi-SearchBox-clear"
        onClick={[Function]}
      >
        Clear
      </button>
      <button
        className="cvi-SearchBox-submit"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;
